# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd
import numpy as np

In [6]:
data_path = r"C:\Users\Вальчук Богдан\OneDrive\Документы\Курс з дата аналітики\Модуль 10. Бібліотека Pandas ключові концепції та методи роботи з даними\Data\supermarket_sales.csv"
df = pd.read_csv(data_path)

In [7]:
type(df)

pandas.core.frame.DataFrame

In [8]:
pd.read_csv(data_path, nrows=100)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
95,594-34-4444,A,Yangon,Normal,Male,Electronic accessories,97.16,1,4.8580,102.0180,3/8/2019,20:38,Ewallet,97.16,4.761905,4.8580,7.2
96,766-85-7061,B,Mandalay,Normal,Male,Health and beauty,87.87,10,43.9350,922.6350,3/29/2019,10:25,Ewallet,878.70,4.761905,43.9350,5.1
97,871-39-9221,C,Naypyitaw,Normal,Female,Electronic accessories,12.45,6,3.7350,78.4350,2/9/2019,13:11,Cash,74.70,4.761905,3.7350,4.1
98,865-92-6136,A,Yangon,Normal,Male,Food and beverages,52.75,3,7.9125,166.1625,3/23/2019,10:16,Ewallet,158.25,4.761905,7.9125,9.3


1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [10]:
df.sort_values(by="Unit price", ascending=False).head()

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
983,148-41-7930,C,Naypyitaw,Normal,Male,Health and beauty,99.96,7,34.986,734.706,1/23/2019,10:33,Cash,699.72,4.761905,34.986,6.1
122,219-22-9386,B,Mandalay,Member,Male,Sports and travel,99.96,9,44.982,944.622,3/9/2019,17:26,Credit card,899.64,4.761905,44.982,4.2
930,641-62-7288,B,Mandalay,Normal,Male,Home and lifestyle,99.92,6,29.976,629.496,3/24/2019,13:33,Ewallet,599.52,4.761905,29.976,7.1
494,437-53-3084,B,Mandalay,Normal,Male,Fashion accessories,99.89,2,9.989,209.769,2/26/2019,11:48,Ewallet,199.78,4.761905,9.989,7.1
283,667-92-0055,A,Yangon,Member,Male,Health and beauty,99.83,6,29.949,628.929,3/4/2019,15:02,Ewallet,598.98,4.761905,29.949,8.5


2. Знайдіть інвойс (рядок в даних) в категорії товарів `Electronic accessories`, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [15]:
df[df['Product line'] == 'Electronic accessories'].sort_values(
    by=['Unit price','Quantity'], ascending=[False,True]).head(1)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating,New col
209,817-69-8206,B,Mandalay,Normal,Female,Electronic accessories,99.73,9,44.8785,942.4485,3/2/2019,19:42,Credit card,897.57,4.761905,44.8785,6.5,Rating Category


3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу (`Rating`): високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 (включно) до 8 (не включно) - то він середній (`Medium`)
- інакше - низький (`Low`).

In [16]:
df['New col'] = "Rating_Category"

In [17]:
df['New col']

0      Rating_Category
1      Rating_Category
2      Rating_Category
3      Rating_Category
4      Rating_Category
            ...       
995    Rating_Category
996    Rating_Category
997    Rating_Category
998    Rating_Category
999    Rating_Category
Name: New col, Length: 1000, dtype: object

In [22]:
df["Rating_Category"] = df['Rating'].apply(lambda x: 'High' if x >= 8 else ( "Medium" if 6 <= x < 8 else 'Low' ))

In [25]:
df[['Rating', 'Rating_Category']]

,Rating,Rating_Category
0,9.1,High
1,9.6,High
2,7.4,Medium
3,8.4,High
4,5.3,Low
...,...,...
995,6.2,Medium
996,4.4,Low
997,7.7,Medium
998,4.1,Low


3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [26]:
df['Rating_Category'].value_counts(normalize=True)*100

Rating_Category
Medium    35.6
High      32.9
Low       31.5
Name: proportion, dtype: float64

4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [27]:
time = '11:28'

In [30]:
def get_hour(t):
    parts = t.split(':')
    hour_str = parts[0] 
    hour_int = int(hour_str)
    return hour_int
    

In [31]:
type(get_hour(time))

int

4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [34]:
df['Hour'] = df['Time'].apply(get_hour)

In [35]:
df.Hour

0      13
1      10
2      13
3      20
4      10
       ..
995    13
996    17
997    13
998    15
999    13
Name: Hour, Length: 1000, dtype: int64

4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [36]:
df['Hour'].value_counts()

Hour
19    113
13    103
15    102
10    101
18     93
11     90
12     89
14     83
16     77
20     75
17     74
Name: count, dtype: int64

Tри найпіковіші години — 19:00 (113), 13:00 (103), 15:00 (102), за ними 10:00 (101).
13:00 (103) - скоріше за все час обідньої перерви
19:00 (113) - робочий день скінчився  - час для закупів
15:00 (102) - кінець першої зміни (наприклад, ранкова зміна з 7:00 чи 8:00 до 15:00) відкриває вікно, коли ці люди по дорозі додому заходять за покупками.
10:00 (101) - початок робочого дня, люди забігають перед роботою чи одразу після відкриття магазину.

Логіка найбільших покупок саме у вказані години видається реальною.


5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [39]:
df.groupby('City')['Total'].mean()

City
Mandalay     319.872506
Naypyitaw    337.099715
Yangon       312.354031
Name: Total, dtype: float64

6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [42]:
df.groupby('City')['Total'].agg(['count','min', 'mean', 'max'])


,count,min,mean,max
City,,,,
Mandalay,332,18.6375,319.872506,1022.49
Naypyitaw,328,10.6785,337.099715,1042.65
Yangon,340,12.6945,312.354031,1039.29


З попереднього дослідження: Yangon мав нижчий gross income (14.87) за Mandalay (15.23), і різниця пояснювалась переважно ціною товарів (Unit price 54.78 vs 55.66, ~1.6%), з невеликим внеском від міксу категорій товарів. Тобто нижчий Yangon там теж підтверджувався даними, не просто випадковість.

Це узгоджується з поточною картиною: у Yangon найбільше покупок (340), але найнижчий середній чек (312.35) — стабільна модель для цього міста в обох дослідженнях. У Naypyitaw купують більш дорогі товари, там найменше покупок (328), але найвищий середній чек (337.10) — протилежна модель поведінки покупця.

Для рекомендації мережі можна об'єднати обидва спостереження: Yangon — стратегія на об'єм (багато дешевших покупок), Naypyitaw — стратегія на чек (менше клієнтів, але кожен витрачає більше). Це може означати різні маркетингові підходи для філій: у Yangon варто працювати над збільшенням середнього чека (крос-продажі, бандли), а в Naypyitaw — над залученням більшої кількості клієнтів (реклама, промо для нових покупців).

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [44]:
df.to_csv('supermarket_sales_enriched.csv', index=False)